core/src/main/java/org/elasticsearch/common/bytes/CompositeBytesReference.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.lucene.util.BytesRefIterator;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+/**
+ * A composite {@link BytesReference} that allows joining multiple bytes references
+ * into one without copying.
+ *
+ * Note, {@link #toBytesRef()} will materialize all pages in this BytesReference.
+ */
+public final class CompositeBytesReference extends BytesReference {
+
+    private final BytesReference[] references;
+    private final int[] offsets;
+    private final int length;
+    private final long ramBytesUsed;
+
+    public CompositeBytesReference(BytesReference... references) {
+        this.references = references;
+        this.offsets = new int[references.length];
+        long ramBytesUsed = 0;
+        int offset = 0;
+        for (int i = 0; i < references.length; i++) {
+            BytesReference reference = references[i];
+            if (reference == null) {
+                throw new IllegalArgumentException("references must not be null");
+            }
+            offsets[i] = offset;
+            offset += reference.length();
+            ramBytesUsed += reference.ramBytesUsed();
+        }
+        this.ramBytesUsed = ramBytesUsed + Integer.BYTES * offsets.length;
+        length = offset;
+    }
+
+
+    @Override
+    public byte get(int index) {
+        int i = getOffsetIndex(index);
+        return references[i].get(index - offsets[i]);
+    }
+
+    @Override
+    public int length() {
+        return length;
+    }
+
+    @Override
+    public BytesReference slice(int from, int length) {
+        final int to = from + length;
+        final int limit = getOffsetIndex(from + length);
+        final int start = getOffsetIndex(from);
+        BytesReference[] inSlice = new BytesReference[1 + (limit - start)];
+        for (int i = 0, j = start; i < inSlice.length; i++) {
+            inSlice[i] = references[j++];
+        }
+        int inSliceOffset = from - offsets[start];
+        if (inSlice.length == 1) {
+            return inSlice[0].slice(inSliceOffset, length);
+        }
+        // now adjust slices in front and at the end
+        inSlice[0] = inSlice[0].slice(inSliceOffset, inSlice[0].length() - inSliceOffset);
+        inSlice[inSlice.length-1] = inSlice[inSlice.length-1].slice(0, to - offsets[limit]);
+        return new CompositeBytesReference(inSlice);
+    }
+
+    private final int getOffsetIndex(int offset) {
+        int i = Arrays.binarySearch(offsets, offset);
+        return i < 0 ? (-(i+1)) - 1 : i;
+    }
+
+    @Override
+    public BytesRef toBytesRef() {
+        BytesRefBuilder builder = new BytesRefBuilder();
+        builder.grow(length());
+        BytesRef spare;
+        BytesRefIterator iterator = iterator();
+        try {
+            while ((spare = iterator.next()) != null) {
+                builder.append(spare);
+            }
+        } catch (IOException ex) {
+            throw new AssertionError("won't happen", ex);
```

Comment:
no but I don't wanna throw that UIO since it's really an error if that happens - we don't do IO in our bytesref iterators?